runners/core-java/src/test/java/org/apache/beam/runners/core/UnboundedReadFromBoundedSourceTest.java

```diff
@@ -0,0 +1,381 @@
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.ReadableByteChannel;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Random;
+
+/**
+ * Unit tests for {@link UnboundedReadFromBoundedSource}.
+ */
+@RunWith(JUnit4.class)
+public class UnboundedReadFromBoundedSourceTest {
+
+  @Rule
+  public TemporaryFolder tmpFolder = new TemporaryFolder();
+
+  @Rule
+  public transient ExpectedException thrown = ExpectedException.none();
+
+  @Test
+  public void testCheckpointCoderNulls() throws Exception {
+    CheckpointCoder<String> coder = new CheckpointCoder<>(StringUtf8Coder.of());
+    Checkpoint<String> emptyCheckpoint = new Checkpoint<>(null, null);
+    Checkpoint<String> decodedEmptyCheckpoint = CoderUtils.decodeFromByteArray(
+        coder,
+        CoderUtils.encodeToByteArray(coder, emptyCheckpoint));
+    assertNull(decodedEmptyCheckpoint.getResidualElements());
+    assertNull(decodedEmptyCheckpoint.getResidualSource());
+  }
+
+  @Test
+  @Category(RunnableOnService.class)
+  public void testBoundedToUnboundedSourceAdapter() throws Exception {
+    long numElements = 100;
+    BoundedSource<Long> boundedSource = CountingSource.upTo(numElements);
+    UnboundedSource<Long, Checkpoint<Long>> unboundedSource =
+        new BoundedToUnboundedSourceAdapter<>(boundedSource);
+
+    Pipeline p = TestPipeline.create();
+
+    PCollection<Long> output =
+        p.apply(Read.from(unboundedSource).withMaxNumRecords(numElements));
+
+    // Count == numElements
+    PAssert
+      .thatSingleton(output.apply("Count", Count.<Long>globally()))
+      .isEqualTo(numElements);
+    // Unique count == numElements
+    PAssert
+      .thatSingleton(output.apply(RemoveDuplicates.<Long>create())
+                          .apply("UniqueCount", Count.<Long>globally()))
+      .isEqualTo(numElements);
+    // Min == 0
+    PAssert
+      .thatSingleton(output.apply("Min", Min.<Long>globally()))
+      .isEqualTo(0L);
+    // Max == numElements-1
+    PAssert
+      .thatSingleton(output.apply("Max", Max.<Long>globally()))
+      .isEqualTo(numElements - 1);
+    p.run();
+  }
+
+  @Test
+  public void testCountingSourceToUnboundedCheckpoint() throws Exception {
+    long numElements = 100;
+    BoundedSource<Long> countingSource = CountingSource.upTo(numElements);
+    List<Long> expected = Lists.newArrayList();
+    for (long i = 0; i < numElements; ++i) {
+      expected.add(i);
+    }
+    testBoundedToUnboundedSourceAdapterCheckpoint(countingSource, expected);
+  }
+
+  @Test
+  public void testCompressedSourceToUnboundedCheckpoint() throws Exception {
+    String baseName = "test-input";
+    File compressedFile = tmpFolder.newFile(baseName + ".gz");
+    byte[] input = generateInput(100);
+    writeFile(compressedFile, input, CompressionMode.GZIP);
+
+    CompressedSource<Byte> source =
+        CompressedSource.from(new ByteSource(compressedFile.getPath(), 1));
```

Comment:
why bring in CompressedSource itself? Why not just add a flag to ByteSource that you can set to make it unsplittable?